tutorial/source/effect_handlers.ipynb

```diff
@@ -0,0 +1,629 @@
+    "\n",
+    "pyro.set_rng_seed(101)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "Inference in probabilistic programming involves manipulating or transforming probabilistic programs written as generative models. For example, nearly all approximate inference algorithms require computing the unnormalized joint probability of values of latent and observed variables under a generative model.\n",
+    "\n",
+    "Consider the following example model from the [introductory inference tutorial](http://pyro.ai/examples/intro_part_ii.html):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def scale(guess):\n",
+    "    weight = pyro.sample(\"weight\", dist.Normal(guess, 1.0))\n",
+    "    return pyro.sample(\"measurement\", dist.Normal(weight, 0.75))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This model defines a joint probability distribution over `\"weight\"` and `\"measurement\"`:\n",
+    "\n",
+    "$${\\sf weight} \\, | \\, {\\sf guess} \\sim \\cal {\\sf Normal}({\\sf guess}, 1) $$\n",
+    "$${\\sf measurement} \\, | \\, {\\sf guess}, {\\sf weight} \\sim {\\sf Normal}({\\sf weight}, 0.75)$$\n",
+    "\n",
+    "If we had access to the inputs and outputs of each `pyro.sample` site, we could compute their log-joint:\n",
+    "```python\n",
+    "logp = dist.Normal(guess, 1.0).log_prob(weight).sum() + dist.Normal(weight, 0.75).log_prob(measurement).sum()\n",
+    "```\n",
+    "However, the way we wrote `scale` above does seem to expose these intermediate distribution objects, and rewriting it to return them would be intrusive and would violate the separation of concerns between models and inference algorithms that a probabilistic programming language like Pyro is designed to enforce.\n",
+    "\n",
+    "To resolve this conflict and facilitate inference algorithm development, Pyro exposes [Poutine](http://docs.pyro.ai/en/dev/poutine.html), a library of [*effect handlers*](http://homepages.inf.ed.ac.uk/slindley/papers/handlers.pdf), or composable building blocks for examining and modifying the behavior of Pyro programs. Most of Pyro's internals are implemented on top of Poutine."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A first look at Poutine: Pyro's library of algorithmic building blocks\n",
+    "\n",
+    "Effect handlers add *side effects* to the behavior of `pyro.sample` and other operations. Before reviewing more definitions, let's look at a first example that addresses the problem above: we can compose two existing effect handlers, `poutine.condition` (which sets output values of `pyro.sample` statements) and `poutine.trace` (which records the inputs, distributions, and outputs of `pyro.sample` statements), to concisely define a new effect handler that computes the log-joint:"
```

Comment:
I think it would be helpful to add a paragraph of background material on algebraic effects, something like

> Algebraic effects are an abstraction common in the programming community. First developed by Gordon Plotkin in 19xx, algebraic effects create abstractions to give nonstandard interpretations to particular statements in a programming language (e.g. the `pyro.sample` and `pyro.param` statements). For an introduction to algebraic effects we recommend the following papers:
    - paper1
    - paper2
    - paper3

Reply:
I added this at the end